File: src/fable/Fable.Core/ts/String.ts
import { toString } from "./Util"
import { getRestParams } from "./Util"
import { escape } from "./RegExp"
import { DateKind } from "./Date"
import { second } from "./Date"
import { minute } from "./Date"
import { hour } from "./Date"
import { day } from "./Date"
import { month } from "./Date"
import { year } from "./Date"

const fsFormatRegExp = /(^|[^%])%([0+ ]*)(-?\d+)?(?:\.(\d+))?(\w)/;
const formatRegExp = /\{(\d+)(,-?\d+)?(?:\:(.+?))?\}/g;

function toHex(value : number) {
  return value < 0
    ? "ff" + (16777215 - (Math.abs(value) - 1)).toString(16)
    : value.toString(16);
}

export function fsFormat(str: string, ...args: any[]): Function | string {
  let _cont: any;
  function isObject(x: any) {
    return x !== null && typeof x === "object" && !(x instanceof Number) && !(x instanceof String) && !(x instanceof Boolean);
  }
  function formatOnce(str: any, rep: any) {
    return str.replace(fsFormatRegExp, function (_: any, prefix: any, flags: any, pad: any, precision: any, format: any) {
      switch (format) {
        case "f": case "F":
          rep = rep.toFixed(precision || 6); break;
        case "g": case "G":
          rep = rep.toPrecision(precision); break;
        case "e": case "E":
          rep = rep.toExponential(precision); break;
        case "O":
          rep = toString(rep); break;
        case "A":
          try {
            rep = JSON.stringify(rep, function (k, v) {
              return v && v[Symbol.iterator] && !Array.isArray(v) && isObject(v) ? Array.from(v)
                : v && typeof v.ToString === "function" ? toString(v) : v;
            });
          }
          catch (err) {
            // Fallback for objects with circular references
            rep = "{" + Object.getOwnPropertyNames(rep).map(k => k + ": " + String(rep[k])).join(", ") + "}";
          }
          break;
        case "x":
          rep = toHex(Number(rep)); break;
        case "X":
          rep = toHex(Number(rep)).toUpperCase(); break;
      }
      const plusPrefix = flags.indexOf("+") >= 0 && parseInt(rep) >= 0;
      if (!isNaN(pad = parseInt(pad))) {
        const ch = pad >= 0 && flags.indexOf("0") >= 0 ? "0" : " ";
        rep = padLeft(rep, Math.abs(pad) - (plusPrefix ? 1 : 0), ch, pad < 0);
      }
      let once = prefix + (plusPrefix ? "+" + rep : rep);
      return once.replace(/%/g, "%%");
    });
  }
  function makeFn(str: any) {
    return (rep: any) => {
      const str2 = formatOnce(str, rep);
      return fsFormatRegExp.test(str2)
        ? makeFn(str2) : _cont(str2.replace(/%%/g, "%"));
    };
  }
  if (args.length === 0) {
    return (cont: any) => {
      _cont = cont;
      return fsFormatRegExp.test(str) ? makeFn(str) : _cont(str);
    };
  }
  else {
    for (let i = 0; i < args.length; i++) {
      str = formatOnce(str, args[i]);
    }
    return str.replace(/%%/g, "%");
  }
}

export function format(str: string, ...args: any[]) {
  return str.replace(formatRegExp, function (match: any, idx: any, pad: any, format: any) {
    let rep = args[idx], padSymbol = " ";
    if (typeof rep === "number") {
      switch ((format || "").substring(0, 1)) {
        case "f": case "F":
          rep = format.length > 1 ? rep.toFixed(format.substring(1)) : rep.toFixed(2);
          break;
        case "g": case "G":
          rep = format.length > 1 ? rep.toPrecision(format.substring(1)) : rep.toPrecision();
          break;
        case "e": case "E":
          rep = format.length > 1 ? rep.toExponential(format.substring(1)) : rep.toExponential();
          break;
        case "p": case "P":
          rep = (format.length > 1 ? (rep * 100).toFixed(format.substring(1)) : (rep * 100).toFixed(2)) + " %";
          break;
        case "x":
         rep = toHex(Number(rep)); break;
        case "X":
          rep = toHex(Number(rep)).toUpperCase(); break;
        default:
          const m = /^(0+)(\.0+)?$/.exec(format);
          if (m != null) {
            let decs = 0;
            if (m[2] != null)
              rep = rep.toFixed(decs = m[2].length - 1);
            pad = "," + (m[1].length + (decs ? decs + 1 : 0)).toString();
            padSymbol = "0";
          } else if (format) {
            rep = format;
          }
      }
    } else if (rep instanceof Date) {
      if (format.length === 1) {
        switch (format) {
          case "D":
            rep = rep.toDateString(); break;
          case "T":
            rep = rep.toLocaleTimeString(); break;
          case "d":
            rep = rep.toLocaleDateString(); break;
          case "t":
            rep = rep.toLocaleTimeString().replace(/:\d\d(?!:)/, ""); break;
          case "o": case "O":
            if ((rep as any).kind === DateKind.Local) {
              const offset = rep.getTimezoneOffset() * -1;
              rep = format("{0:yyyy-MM-dd}T{0:HH:mm}:{1:00.000}{2}{3:00}:{4:00}",
                                    rep, second(rep), offset >= 0 ? "+" : "-",
                                    ~~(offset / 60), offset % 60);
            }
            else {
              rep = rep.toISOString()
            }
        }
      } else {
        rep = format.replace(/\w+/g, function (match2: any) {
          let rep2 = match2;
          switch (match2.substring(0, 1)) {
            case "y":
              rep2 = match2.length < 4 ? year(rep) % 100 : year(rep);
              break;
            case "h":
              rep2 = rep.getHours() > 12 ? hour(rep) % 12 : hour(rep);
              break;
            case "M":
              rep2 = month(rep);
              break;
            case "d":
              rep2 = day(rep);
              break;
            case "H":
              rep2 = hour(rep);
              break;
            case "m":
              rep2 = minute(rep);
              break;
            case "s":
              rep2 = second(rep);
              break;
          }
          if (rep2 !== match2 && rep2 < 10 && match2.length > 1) {
            rep2 = "0" + rep2;
          }
          return rep2;
        });
      }
    }
    if (!isNaN(pad = parseInt((pad || "").substring(1)))) {
      rep = padLeft(rep, Math.abs(pad), padSymbol, pad < 0);
    }
    return rep;
  });
}

export function endsWith(str: string, search: string) {
  const idx = str.lastIndexOf(search);
  return idx >= 0 && idx == str.length - search.length;
}

export function initialize(n: number, f: (i: number) => string) {
  if (n < 0)
    throw new Error("String length must be non-negative");
  const xs = new Array(n);
  for (let i = 0; i < n; i++)
    xs[i] = f(i);
  return xs.join("");
}

export function insert(str: string, startIndex: number, value: string) {
  if (startIndex < 0 || startIndex > str.length) {
    throw new Error("startIndex is negative or greater than the length of this instance.");
  }
  return str.substring(0, startIndex) + value + str.substring(startIndex);
}

export function isNullOrEmpty(str: string | any) {
  return typeof str !== "string" || str.length == 0;
}

export function isNullOrWhiteSpace(str: string | any) {
  return typeof str !== "string" || /^\s*$/.test(str);
}

export function join(delimiter: string, xs: ArrayLike<string>) {
  xs = typeof xs == "string" ? getRestParams(arguments, 1) : xs;
  return (Array.isArray(xs) ? xs : Array.from(xs)).join(delimiter);
}

export function newGuid() {
  let uuid = "";
  for (let i = 0; i < 32; i++) {
    const random = Math.random() * 16 | 0;
    if (i === 8 || i === 12 || i === 16 || i === 20)
      uuid += "-";
    uuid += (i === 12 ? 4 : i === 16 ? random & 3 | 8 : random).toString(16);
  }
  return uuid;
}

export function padLeft(str: any, len: number, ch?: string, isRight?: boolean) {
  ch = ch || " ";
  str = String(str);
  len = len - str.length;
  for (let i = -1; ++i < len;)
    str = isRight ? str + ch : ch + str;
  return str;
}

export function padRight(str: any, len: number, ch?: string) {
  return padLeft(str, len, ch, true);
}

export function remove(str: string, startIndex: number, count?: number) {
  if (startIndex >= str.length) {
    throw new Error("startIndex must be less than length of string");
  }
  if (typeof count === "number" && (startIndex + count) > str.length) {
    throw new Error("Index and count must refer to a location within the string.")
  }
  return str.slice(0, startIndex) + (typeof count === "number" ? str.substr(startIndex + count) : "");
}


export function replace(str: string, search: string, replace: string) {
  return str.replace(new RegExp(escape(search), "g"), replace);
}

export function replicate(n: number, x: string) {
  return initialize(n, () => x);
}

export function split(str: string, splitters: string[], count?: number, removeEmpty?: number) {
  count = typeof count == "number" ? count : null;
  removeEmpty = typeof removeEmpty == "number" ? removeEmpty : null;
  if (count < 0)
    throw new Error("Count cannot be less than zero");
  if (count === 0)
    return [];
  splitters = Array.isArray(splitters) ? splitters : getRestParams(arguments, 1);
  splitters = splitters.map(x => escape(x));
  splitters = splitters.length > 0 ? splitters : [" "];
  let m: RegExpExecArray;
  let i = 0;
  const splits: string[] = [];
  const reg = new RegExp(splitters.join("|"), "g");
  while ((count == null || count > 1) && (m = reg.exec(str)) !== null) {
    if (!removeEmpty || (m.index - i) > 0) {
      count = count != null ? count - 1 : count;
      splits.push(str.substring(i, m.index));
    }
    i = reg.lastIndex;
  }
  if (!removeEmpty || (str.length - i) > 0)
    splits.push(str.substring(i));
  return splits;
}

export function trim(str: string, side: "start" | "end" | "both", ...chars: string[]) {
  if (side == "both" && chars.length == 0)
    return str.trim();
  if (side == "start" || side == "both") {
    const reg = chars.length == 0 ? /^\s+/ : new RegExp("^[" + escape(chars.join("")) + "]+");
    str = str.replace(reg, "");
  }
  if (side == "end" || side == "both") {
    const reg = chars.length == 0 ? /\s+$/ : new RegExp("[" + escape(chars.join("")) + "]+$");
    str = str.replace(reg, "");
  }
  return str;
}

File: src/fable-library-py/fable_library/string.py
import locale
import re

from base64 import b64decode, b64encode
from dataclasses import dataclass
from datetime import datetime
from enum import IntEnum
from typing import (
    Any,
    Callable,
    Iterable,
    List,
    Match,
    NoReturn,
    Optional,
    Pattern,
    TypeVar,
    Union,
    cast,
    overload,
)

from .date import to_string as date_to_string
from .numeric import multiply, to_exponential, to_fixed, to_hex, to_precision
from .reg_exp import escape
from .types import to_string


T = TypeVar("T")


_fs_format_regexp: Pattern[str] = re.compile(
    r"(^|[^%])%([0+\- ]*)(\d+)?(?:\.(\d+))?(\w)"
)
_interpolate_regexp: Pattern[str] = re.compile(
    r"(?:(^|[^%])%([0+\- ]*)(\d+)?(?:\.(\d+))?(\w))?%P\(\)"
)
_format_regexp: Pattern[str] = re.compile(
    r"\{(\d+)(,-?\d+)?(?:\:([a-zA-Z])(\d{0,2})|\:(.+?))?\}"
)


IPrintfFormatContinuation = Callable[[Callable[[str], Any]], Callable[[str], Any]]


@dataclass
class IPrintfFormat:
    input: str
    cont: IPrintfFormatContinuation


def printf(input: str) -> IPrintfFormat:
    format: IPrintfFormatContinuation = fs_format(input)
    return IPrintfFormat(input=input, cont=format)


def continue_print(cont: Callable[[str], Any], arg: Union[IPrintfFormat, str]) -> Any:
    """Print continuation."""

    if isinstance(arg, IPrintfFormat):
        ret = arg.cont(cont)
        return ret

    return cont(arg)


def to_console(arg: Union[IPrintfFormat, Any]) -> Any:
    return continue_print(print, arg)


def to_console_error(arg: Union[IPrintfFormat, str]):
    return continue_print(lambda x: print(x), arg)


# Set return to `Any` since `Union[str, Callable]` will make type
# checkers really unhappy.
def to_text(arg: Union[IPrintfFormat, str]) -> Union[str, Callable[..., Any]]:
    cont: Callable[[str], Any] = lambda x: x
    return continue_print(cont, arg)


def to_fail(arg: Union[IPrintfFormat, str]):
    def fail(msg: str):
        raise Exception(msg)

    return continue_print(fail, arg)


def format_replacement(
    rep: Any, flags: Any, padLength: Any, precision: Any, format: Any
):
    sign = ""
    flags = flags or ""
    format = format or ""

    if isinstance(rep, (int, float)):
        if format not in ["x", "X"]:
            if rep < 0:
                rep = rep * -1
                sign = "-"
            else:
                if flags.find(" ") >= 0:
                    sign = " "
                elif flags.find("+") >= 0:
                    sign = "+"

        if format == "x":
            rep = to_hex(cast(int, rep))
        elif format == "X":
            rep = to_hex(cast(int, rep)).upper()
        if format in ("f", "F"):
            precision = int(precision) if precision is not None else 6
            rep = to_fixed(rep, precision)
        elif format in ("g", "G"):
            rep = (
                to_precision(rep, precision)
                if precision is not None
                else to_precision(rep)
            )
        elif format in ("e", "E"):
            rep = (
                to_exponential(rep, precision)
                if precision is not None
                else to_exponential(rep)
            )
        else:  # AOid
            rep = to_string(rep)

    elif isinstance(rep, datetime):
        rep = date_to_string(rep)
    else:
        rep = to_string(rep)

    if padLength is not None:
        padLength = int(padLength)
        zeroFlag = flags.find("0") >= 0  # Use '0' for left padding
        minusFlag = flags.find("-") >= 0  # Right padding
        ch = " " if minusFlag or not zeroFlag else "0"
        if ch == "0":
            rep = pad_left(rep, padLength - len(sign), ch, minusFlag)
            rep = sign + rep
        else:
            rep = pad_left(sign + rep, padLength, ch, minusFlag)

    else:
        rep = sign + rep

    return rep


def interpolate(string: str, values: List[Any]) -> str:
    valIdx = 0
    strIdx = 0
    result = ""
    matches = _interpolate_regexp.finditer(string)
    for match in matches:
        # The first group corresponds to the no-escape char (^|[^%]), the actual pattern starts in the next char
        # Note: we don't use negative lookbehind because some browsers don't support it yet
        matchIndex = match.start() + len(match[1] or "")
        result += string[strIdx:matchIndex].replace("%%", "%")
        [_, flags, padLength, precision, format] = match.groups()
        # print(match.groups())
        result += format_replacement(
            values[valIdx], flags, padLength, precision, format
        )
        valIdx += 1

        strIdx = match.end()

    result += string[strIdx:].replace("%%", "%")
    return result


def format_once(str2: str, rep: Any):
    def match(m: Match[str]):
        prefix, flags, pad_length, precision, format = m.groups()
        once: str = format_replacement(rep, flags, pad_length, precision, format)
        return prefix + once.replace("%", "%%")

    ret = _fs_format_regexp.sub(match, str2, count=1)
    return ret


def create_printer(string: str, cont: Callable[..., Any]):
    def _(*args: Any):
        str_copy: str = string
        for arg in args:
            str_copy = format_once(str_copy, arg)

        if _fs_format_regexp.search(str_copy):
            return create_printer(str_copy, cont)
        return cont(str_copy.replace("%%", "%"))

    return _


def fs_format(str: str) -> Callable[[Callable[..., Any]], Any]:
    def _(cont: Callable[..., Any]):
        if _fs_format_regexp.search(str):
            return create_printer(str, cont)
        return cont(str)

    return _


def format(string: str, *args: Any) -> str:
    if not string and args:
        # Called with culture info
        string = args[0]
        args = args[1:]

    def match(m: Match[str]) -> str:
        idx, padLength, format, precision_, pattern = list(m.groups())
        rep = args[int(idx)]
        if isinstance(rep, (int, float)):
            precision: Optional[int] = None
            try:
                precision: Optional[int] = int(precision_)
            except Exception:
                pass

            if format in ["f", "F"]:
                precision = precision if precision is not None else 2
                rep = to_fixed(rep, precision)

            elif format in ["g", "G"]:
                rep = (
                    to_precision(rep, precision)
                    if precision is not None
                    else to_precision(rep)
                )

            elif format in ["e", "E"]:
                rep = (
                    to_exponential(rep, precision)
                    if precision is not None
                    else to_exponential(rep)
                )

            elif format in ["p", "P"]:
                precision = precision if precision is not None else 2
                rep = to_fixed(multiply(rep, 100), precision) + " %"

            elif format in ["d", "D"]:
                rep = (
                    pad_left(str(rep), precision, "0")
                    if precision is not None
                    else str(rep)
                )

            elif format in ["x", "X"] and isinstance(rep, int):
                rep = (
                    pad_left(to_hex(rep), precision, "0")
                    if precision is not None
                    else to_hex(rep)
                )
                if format == "X":
                    rep = rep.upper()
            elif pattern:
                sign = ""

                def match(m: Match[str]):
                    nonlocal sign, rep

                    intPart, decimalPart = list(m.groups())
                    if rep < 0:
                        rep = multiply(rep, -1)
                        sign = "-"

                    rep = to_fixed(rep, len(decimalPart) - 1 if decimalPart else 0)
                    return pad_left(
                        rep,
                        len(intPart or "")
                        - len(sign)
                        + (len(decimalPart) if decimalPart else 0),
                        "0",
                    )

                rep = re.sub(r"(0+)(\.0+)?", match, pattern)
                rep = sign + rep

        elif isinstance(rep, datetime):
            rep = date_to_string(rep, pattern or format)
        else:
            rep = to_string(rep)

        try:
            padLength = int((padLength or " ")[1:])
            rep = pad_left(str(rep), abs(padLength), " ", padLength < 0)
        except ValueError:
            pass

        return str(rep)

    return _format_regexp.sub(match, string)


def initialize(n: int, f: Callable[[int], str]) -> str:
    if n < 0:
        raise Exception("String length must be non-negative")

    return "".join([f(i) for i in range(n)])


def insert(string: str, startIndex: int, value: str):
    if startIndex < 0 or startIndex > len(string):
        raise ValueError(
            "startIndex is negative or greater than the length of this instance."
        )

    return string[:startIndex] + value + string[startIndex:]


def is_null_or_empty(string: Optional[str]):
    return not isinstance(string, str) or not len(string)


def is_null_or_white_space(string: Optional[Any]) -> bool:
    return not string or not isinstance(string, str) or string.isspace()


def concat(*xs: Iterable[Any]) -> str:
    return "".join(map(str, xs))


def join(delimiter: str, xs: Iterable[Any]) -> str:
    return delimiter.join((str(x) for x in xs))


def join_with_indices(delimiter: str, xs: List[str], startIndex: int, count: int):
    endIndexPlusOne = startIndex + count
    if endIndexPlusOne > len(xs):
        raise ValueError("Index and count must refer to a location within the buffer.")

    return delimiter.join(xs[startIndex:endIndexPlusOne])


def not_supported(name: str) -> NoReturn:
    raise Exception(
        "The environment doesn't support '" + name + "', please use a polyfill."
    )


def to_base64string(in_array: bytes) -> str:
    return b64encode(in_array).decode("utf8")


def from_base64string(b64encoded: str) -> bytes:
    return b64decode(b64encoded)


def pad_left(
    string: str, length: int, ch: Optional[str] = None, isRight: Optional[bool] = False
) -> str:
    ch = ch or " "
    length = length - len(string)
    for _ in range(length):
        string = string + ch if isRight else ch + string

    return string


def pad_right(string: str, len: int, ch: Optional[str] = None) -> str:
    return pad_left(string, len, ch, True)


def remove(string: str, startIndex: int, count: Optional[int] = None):
    if startIndex >= len(string):
        raise ValueError("startIndex must be less than length of string")

    if count and (startIndex + count) > len(string):
        raise ValueError("Index and count must refer to a location within the string.")

    return string[:startIndex] + (
        string[startIndex + count :] if count is not None else ""
    )


def replace(string: str, search: str, replace: str):
    return string.replace(search, replace)


def replicate(n: int, x: str) -> str:
    return initialize(n, lambda _=0: x)


def get_char_at_index(input: str, index: int):
    if index < 0 or index >= len(input):
        raise ValueError("Index was outside the bounds of the array.")

    return input[index]


def split(
    string: str,
    splitters: Union[str, List[str]],
    count: Optional[int] = None,
    removeEmpty: int = 0,
) -> List[str]:
    """Split string

    Returns a string array that contains the substrings in this instance
    that are delimited by elements of a specified string or Unicode
    character array."""

    if count and count < 0:
        raise ValueError("Count cannot be less than zero")

    if count == 0:
        return []

    if isinstance(splitters, str):
        if not removeEmpty:
            return string.split(splitters, count - 1 if count else -1)

        splitters = [splitters]

    splitters = [escape(x) for x in splitters] or [" "]

    i = 0
    splits: List[str] = []
    matches = re.finditer("|".join(splitters), string)
    for m in matches:
        if count is not None and count <= 1:
            break

        split = string[i : m.start()]
        if split or not removeEmpty:
            splits.append(split)

            count = count - 1 if count is not None else count

        i = m.end()

    if (count is None or count and count > 0) and len(string) - i > -1:
        split = string[i:]
        if split or not removeEmpty:
            splits.append(split)

    return splits


def trim(string: str, *chars: str) -> str:
    if not len(chars):
        return string.strip()

    pattern = "[" + escape("".join(chars)) + "]+"
    return re.sub(pattern + "$", "", re.sub("^" + pattern, "", string))


def trim_start(string: str, *chars: str) -> str:
    if not len(chars):
        return string.lstrip()

    return re.sub("^[" + escape("".join(chars)) + "]+", "", string)


def trim_end(string: str, *chars: str) -> str:
    if not len(chars):
        return string.rstrip()

    return re.sub("[" + escape("".join(chars)) + "]+$", "", string)


def filter(pred: Callable[[str], bool], x: str) -> str:
    return "".join(c for c in x if pred(c))


def substring(string: str, startIndex: int, length: Optional[int] = None) -> str:
    if startIndex + (length or 0) > len(string):
        raise ValueError("Invalid startIndex and/or length")

    if length is not None:
        return string[startIndex : startIndex + length]

    return string[startIndex:]


class StringComparison(IntEnum):
    CurrentCulture = 0
    CurrentCultureIgnoreCase = 1
    InvariantCulture = 2
    InvariantCultureIgnoreCase = 3
    Ordinal = 4
    OrdinalIgnoreCase = 5


def cmp(x: str, y: str, ic: Union[bool, StringComparison]) -> int:
    def is_ignore_case(i: Union[bool, StringComparison]) -> bool:
        return (
            i is True
            or i == StringComparison.CurrentCultureIgnoreCase
            or i == StringComparison.InvariantCultureIgnoreCase
            or i == StringComparison.OrdinalIgnoreCase
        )

    def is_ordinal(i: Union[bool, int]) -> bool:
        return i == StringComparison.Ordinal or i == StringComparison.OrdinalIgnoreCase

    if not x:
        return 0 if not y else -1
    if not y:
        return 1  # everything is bigger than None

    if is_ordinal(ic):
        if is_ignore_case(ic):
            x = x.lower()
            y = y.lower()

        return 0 if x == y else -1 if x < y else 1
    elif is_ignore_case(ic):
        x = x.lower()
        y = y.lower()

    return locale.strcoll(x, y)


@overload
def compare(string1: str, string2: str, /) -> int:
    """Compares two specified String objects and returns an integer that
    indicates their relative position in the sort order."""
    ...


@overload
def compare(
    string1: str, string2: str, ignore_case: bool, culture: StringComparison, /
) -> int:
    ...


def compare(*args: Any) -> int:
    """Compares two specified String objects and returns an integer that
    indicates their relative position in the sort order.

    All overloads of the Compare method return a 32-bit signed integer
    indicating the lexical relationship between the two comparands.

    - Less than zero: The first substring precedes the second
        substring in the sort order.
    - Zero: The substrings occur in the same position in the sort
        order, or length is zero.
    - Greater than zero: The first substring follows the second
        substring in the sort order.

    Returns:
        Integer that indicates the relationship of the two substrings
        to each other in the sort order:
    """
    length = len(args)

    if length == 2:
        return cmp(args[0], args[1], False)
    if length == 3:
        return cmp(args[0], args[1], args[2])
    if length == 4:
        return cmp(args[0], args[1], args[2])
    if length == 5:
        return cmp(
            args[0][args[1] : args[4] + 1], args[2][args[3] : args[4] + 1], False
        )
    if length == 6:
        return cmp(
            args[0][args[1] : args[4] + 1], args[2][args[3] : args[4] + 1], args[5]
        )
    if length == 7:
        return cmp(
            args[0][args[1] : args[4] + 1], args[2][args[3] : args[4] + 1], args[5]
        )
    raise Exception("String.compare: Unsupported number of parameters")


def compare_to(this: str, other: str) -> int:
    """Compare this string with other

    Compares this instance with a specified String object and indicates
    whether this instance precedes, follows, or appears in the same
    position in the sort order as the specified string.
    """
    return cmp(this, other, StringComparison.CurrentCulture)


def ends_with(string: str, search: str):
    idx = string.rfind(search)
    return idx >= 0 and idx == len(string) - len(search)


def starts_with(string: str, pattern: str, ic: int):
    if len(string) >= len(pattern):
        return cmp(string[0 : len(pattern)], pattern, True if ic else False) == 0

    return False


def index_of_any(string: str, any_of: List[str], *args: int):
    if not string:
        return -1

    start_index = args[0] if len(args) > 0 else 0
    if start_index < 0:
        raise ValueError("Start index cannot be negative")

    length = args[1] if len(args) > 1 else len(string) - start_index
    if length < 0:
        raise ValueError("Length cannot be negative")

    if length > len(string) - start_index:
        raise ValueError("Invalid start_index and length")

    string = string[start_index:length]
    for c in any_of:
        index = string.find(c)
        if index > -1:
            return index + start_index

    return -1
